sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -942,9 +947,32 @@ Scheduler::GraphBuilder::addCG(std::unique_ptr<detail::CG> CommandGroup,
           NeedMemMoveToHost = true;
           MemMoveTargetQueue = HT.MQueue;
         }
-      } else if (!Queue->is_host() && !Record->MCurContext->is_host())
-        NeedMemMoveToHost = true;
-
+      } else if (!Queue->is_host() && !Record->MCurContext->is_host()) {
+        if (detail::getImplBackend(Queue) !=
+            detail::getImplBackend(Record->MCurContext))
+          NeedMemMoveToHost = true;
+        else {
+          std::vector<RT::PiDevice> devs;
+
+          Queue->getPlugin().call_nocheck<PiApiKind::piDeviceGetInfo>(
+              Queue->getDeviceImplPtr()->getHandleRef(),
+              PI_DEVICE_INFO_P2P_READ_DEVICES, sizeof(devs), &devs, nullptr);
+
+          bool can_read_peer = false;
+          const auto &src_dev = findAllocaCmd(Record)
+                                    ->getQueue()
+                                    ->getDeviceImplPtr()
+                                    ->getHandleRef();
+          for (const auto &dev : devs) {
+            if (dev == src_dev) {
+              can_read_peer = true;
+              break;
+            }
+          }
+          if (!can_read_peer)
+            NeedMemMoveToHost = true;
```

Comment:
Seems like this could be simplified to
```c++
NeedMemMoveToHost = std::find(devs.begin(), devs.end(), src_dev) != devs.end();
```
This removes the need for `can_read_peer` as well as the loop above.